sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ProgressReporter.java

```diff
@@ -146,8 +146,22 @@ public void rewindProgress() {
         }
     }
 
-    static Flux<ByteBuffer> addParallelProgressReporting(Flux<ByteBuffer> data, IProgressReceiver progressReceiver,
-        Lock lock, AtomicLong totalProgress) {
+    /**
+     * Adds parallel progress reporting functionality to the given {@code Flux}. Each subscription (and therefore each
+     * retry) will rewind the progress reported so as not to over-report. The data reported will be the total amount
+     * of data emitted so far, or the "current position" of the Flux in parallel.
+     *
+     * @param data The data whose transfer progress is to be tracked.
+     * @param progressReceiver {@link ProgressReceiver}
+     * @param lock This lock will be instantiated by the operation initiating the whole transfer to coordinate each
+     * ProgressReporterImpl.
+     * @param totalProgress We need an AtomicLong to be able to update the value referenced. Because we are already
+     * synchronizing with the lock, we don't incur any additional performance hit here by the synchronization.
+     * @return A {@code Flux} that emits the same data as the source but calls a callback to report the total amount of
+     * data emitted so far.
+     */
+    public static Flux<ByteBuffer> addParallelProgressReporting(Flux<ByteBuffer> data,
```

Comment:
I don't think we want this method public. It's really for our own use when we handle parallel uploads

Reply:
we are using from `com.azure.storage.blob.specialized` (BlobAsyncClientBase#download) and from `com.azure.storage.blob` (BlobAsyncClient:upload).

We would need to make some extra refactoring to have `BlobAsyncClientBase` in same package